### DISCUSSION TITLE
Rendering different component in a shared layout based on page.

### DISCUSSION THREAD

**Discussion by nickythorne** [vercel/next.js#63341] [App Router]
I want to utilise the features of nested layouts into order to have a shared header for child routes. However, I would like a way to inject custom components into that shared layout from the child pages _without having to create a separate layout_ (if possible).

The anatomy of my page structure is as follows:
<img width="1047" alt="Screenshot 2024-03-15 at 21 19 57" src="https://github.com/vercel/next.js/assets/860710/32ea29f8-cad2-4e06-afc8-ec7785a5ad6c">

Is there anyway I can inject custom components from the page component (green) into the shared project layout (red) without having to define a new layout at the page level? The layout is mostly shared between the child routes, but I would like to get the benefits of having the shared layout at project level, but have the ability to inject custom component into that layout for the purposes of say, breadcrumbs - without having to define a new layout every time for the nested pages.

For example, another child layout would be a project settings, but I would like to use the same shared layout, but only change the breadcrumbs.

<img width="1061" alt="Screenshot 2024-03-15 at 21 23 20" src="https://github.com/vercel/next.js/assets/860710/fa15ff04-b006-4977-b757-6ecfeba68494">

**Comment by rettgerst** [comment]
this discussion is marked as resolved with no solution provided.

I'm having effectively this same issue in my application, where I want to place a page title (per-page data) in the same header component as an account selector (layout). as far as I'm aware this isn't possible in next.js' app router without including the header in the page rather than the layout.

@nickythorne did you find a solution to this problem?

**Reply by nickythorne** [reply]
Unfortunately not, I stuck with having to render the page header (which contains page title) in the page itself. I closed the issue because I think this is the intended solution given layouts do not re-render on page changes by design. So there is no opportunity to change the anything in the layout on change of page.

The solutions are to have a layout per page which renders the title, or render things in the page itself.

There is a nice approach to rendering breadcrumbs dynamically however using a combination of slots and variadic path segments.

You would do something like: `@ breadcrumbs/[…paths]/page.tsx` and in your layout:

```
function Layout({ breadcrumbs, children }) {
return ( 
   <main>
     {breadcrumbs}
     {children}
   </main>
 )}
```
From the breadcrumbs page, the paths params will contains all the routes segments, so you could produce the list of breadcrumbs dynamically based on this. 

I guess a similar approach could be taken in your case for rendering the page title based on some match of the path segments and use a slot to render the account selector?

Obviously far from ideal as the page title logic is separate from the page itself, but there is no other way to "inject" dynamic data from the child page into a parent layout, which as I understand, is by design.
